### ISSUE TITLE
HTTP2: Http2Connection.SendHeadersAsync doesn't pass its cancellation token to StartWriteAsync

### ISSUE THREAD

**Issue by marklio** [dotnet/corefx#39634]
I experienced this as a hang under load on the client during gRPC stress. I wanted to get this filed before the weekend while I still had the state saved. It looks like this is just an oversight. I'll circle back with a PR.

This is likely to block continued reliability testing in this area.

**Comment by geoffkizer** [comment]
Are you actually hitting issues with this? In general, the writer lock shouldn't be held for long. The only case in which it might be held for a long time is if the underlying socket send buffer is full, which is probably indicative of more serious problems, likely on the server side.

I do think we should fix this, just trying to understand the urgency. Also, there's a related known issue with cancellation here -- #39240. That's currently punted from 3.0. If you're hitting issues because the socket send buffer is full, you'll probably hit issues with that too.

**Comment by marklio** [comment]
> Are you actually hitting issues with this? In general, the writer lock shouldn't be held for long. The only case in which it might be held for a long time is if the underlying socket send buffer is full, which is probably indicative of more serious problems, likely on the server side.

Yes. I'm happy to share the dump with you on Monday. Below is !dumpasync -stacks output to give you an idea what's going on. In the "long" async stack, we're waiting on the lock and the Available cancellation token would break us out of that wait had it been passed. This particular dump also reveals that SendRstStreamAsync also calls StartWriteAsync without a CancellationToken and has what looks like similar hangs. In this case, killing the server did not break us out of the hangs (in fact, the process is still running, hung). It's definitely possible there's something else at the root here.

Now that we know we can hit these kinds of issues with fairly vanilla gRPC calls, I'm worried if we're punting the issues we know about for 3.0.

```
0:000> !dumpasync -stacks
Statistics:
              MT    Count    TotalSize Class Name
00007ffa26201f80        1           96 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[GrpcGreeterClient.Program+<BasicStress>d__5, GrpcGreeterClient]]
00007ffa261c0ab0        1           96 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[GrpcGreeterClient.Program+<>c__DisplayClass5_0+<<BasicStress>g__Request|0>d, GrpcGreeterClient]]
00007ffa261bf1d0        1          136 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[Grpc.Core.Metadata, Grpc.Core.Api],[Grpc.Net.Client.Internal.GrpcCall`2+<GetResponseHeadersAsync>d__64[[GrpcGreeter.HelloRequest, GrpcGreeterClient],[GrpcGreeter.HelloReply, GrpcGreeterClient]], Grpc.Net.Client]]
00007ffa261b4c50        1          144 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[Grpc.Net.Client.Internal.GrpcCall`2+<SendAsync>d__73[[GrpcGreeter.HelloRequest, GrpcGreeterClient],[GrpcGreeter.HelloReply, GrpcGreeterClient]], Grpc.Net.Client]]
00007ffa2619f190        1          144 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Net.Http.HttpResponseMessage, System.Net.Http],[System.Net.Http.RedirectHandler+<SendAsync>d__4, System.Net.Http]]
00007ffa261b23a0        1          152 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Net.Http.HttpResponseMessage, System.Net.Http],[System.Net.Http.HttpClient+<FinishSendAsyncUnbuffered>d__71, System.Net.Http]]
00007ffa261bbe20        1          168 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[GrpcGreeter.HelloReply, GrpcGreeterClient],[Grpc.Net.Client.Internal.GrpcCall`2+<GetResponseAsync>d__66[[GrpcGreeter.HelloRequest, GrpcGreeterClient],[GrpcGreeter.HelloReply, GrpcGreeterClient]], Grpc.Net.Client]]
00007ffa2619e590        1          184 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Net.Http.HttpResponseMessage, System.Net.Http],[System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__47, System.Net.Http]]
00007ffa261ccc80        1          216 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Net.Http.HttpResponseMessage, System.Net.Http],[System.Net.Http.Http2Connection+<SendAsync>d__79, System.Net.Http]]
00007ffa261cc250        1          224 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Net.Http.Http2Connection+Http2Stream, System.Net.Http],[System.Net.Http.Http2Connection+<SendHeadersAsync>d__64, System.Net.Http]]
00007ffa2624b0b8       18         2304 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d, System.Net.Http]]
00007ffa2624acc8       18         2448 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54, System.Net.Http]]
00007ffa261e7478       19         2584 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[System.Net.Http.Http2Connection+<StartWriteAsync>d__49, System.Net.Http]]
00007ffa261e7028       19         2584 System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1+AsyncStateMachineBox`1[[System.Threading.Tasks.VoidTaskResult, System.Private.CoreLib],[System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51, System.Net.Http]]
Total 84 objects
In 19 chains.
         Address               MT     Size      State Description
00000212a332b350 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332b3d8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332b460 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332b4e8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332b678 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332b700 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a32ae668 (2) System.Net.Http.Http2Connection+<SendHeadersAsync>d__64
...00000212a32ae748 (0) System.Net.Http.Http2Connection+<SendAsync>d__79
....00000212a32ae820 (2) System.Net.Http.HttpConnectionPool+<SendWithRetryAsync>d__47
.....00000212a32ae8d8 (0) System.Net.Http.RedirectHandler+<SendAsync>d__4
......00000212a32ae968 (0) System.Net.Http.HttpClient+<FinishSendAsyncUnbuffered>d__71
.......00000212a32aea00 (0) Grpc.Net.Client.Internal.GrpcCall`2+<SendAsync>d__73[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]]
........00000212a32aea90 (0) Grpc.Net.Client.Internal.GrpcCall`2+<GetResponseAsync>d__66[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]]
........00000212a32aeb38 (0) Grpc.Net.Client.Internal.GrpcCall`2+<GetResponseHeadersAsync>d__64[[System.__Canon, System.Private.CoreLib],[System.__Canon, System.Private.CoreLib]]
.........00000212a32aec18 (0) GrpcGreeterClient.Program+<>c__DisplayClass5_0+<<BasicStress>g__Request|0>d
..........00000212a331dfb8 (1) GrpcGreeterClient.Program+<BasicStress>d__5
...........00000212a331e018 System.Threading.Tasks.Task+SetOnInvokeMres
--------------------------------------------------------------------------------
00000212a332b7e0 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332b868 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332b8f0 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332b978 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332ba50 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332bad8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332bb60 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332bbe8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332c048 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332c0d0 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332c158 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332c1e0 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332c350 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332c3d8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332c460 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332c4e8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332c850 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332c8d8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332c960 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332c9e8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332d918 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332d9a0 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332da28 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332dab0 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332dc20 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332dca8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332dd30 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a332ddb8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a332fe70 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a332fef8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a332ff80 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3330008 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3330410 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3330498 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3330520 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a33305a8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3330680 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3330708 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3330790 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3330818 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3330e10 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3330e98 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3330f20 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3330fa8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3331080 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3331108 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3331190 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3331218 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3331aa0 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3331b28 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3331bb0 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3331c38 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3333538 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a33335c0 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3333648 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a33336d0 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3333d60 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3333de8 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3333e70 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a3333ef8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3333fd0 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a3334058 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3334870 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a33348f8 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------
00000212a3334138 00007ffa261e7028      136          0 System.Net.Http.Http2Connection+<AcquireWriteLockAsync>d__51
Async "stack":
.00000212a33341c0 (0) System.Net.Http.Http2Connection+<StartWriteAsync>d__49
..00000212a3334248 (0) System.Net.Http.Http2Connection+<SendRstStreamAsync>d__54
...00000212a33342d0 (0) System.Net.Http.HttpConnectionBase+<<IgnoreExceptions>g__IgnoreExceptionsAsync|7_0>d
--------------------------------------------------------------------------------

```

**Comment by geoffkizer** [comment]
Thanks for the info.

The interesting question here is why the lock isn't getting released. Since you killed the server, it doesn't sound like it's a server/send buffer issue. Most likely we are hitting some error and neglecting to release the lock somewhere, which is really bad.

@stephentoub suggestions on how to track this down?

**Comment by geoffkizer** [comment]
One thing I noticed and fixed in my current PR is that we're not releasing the writer lock correctly when we think the connection is disposed: https://github.com/dotnet/corefx/blob/3c71b23c676c3538b33f4ae653cf6f85de77e44d/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L809

It's possible you could be hitting this, though I'm not clear why we would have gotten in to this state.

You could try grabbing my PR, building, and testing with that. Or I could create a separate PR just for that issue, to minimize other changes.

**Comment by geoffkizer** [comment]
> SendRstStreamAsync also calls StartWriteAsync without a CancellationToken and has what looks like similar hangs

This is not a problem. The SendRstStreamAsync calls are not awaited and won't block request completion. Basically, this means we have a bunch of RST_STREAM frames queued up to send.

**Comment by marklio** [comment]
> The interesting question here is why the lock isn't getting released. Since you killed the server, it doesn't sound like it's a server/send buffer issue. Most likely we are hitting some error and neglecting to release the lock somewhere, which is really bad.

Just wanted to point out that this seems to be triggered in a cancellation scenario. Killing the server happened hours after the hang.

**Comment by karelz** [comment]
This bug was fixed in 3.0 in PR #39646
